Translation mixins considered harmful (#425)
reactjs/pt-BR.reactjs.org: content/blog/2016-07-13-mixins-considered-harmful.md

---

> titulo: "Considerando Mixins Nocivos"
> autor: [gaearon]

Reviewer: O erro no build é porque vc traduziu as tags do markdown, essas partes não pode traduzir, vc pode ver os erros no log do netlify

```suggestion
title: "Considerando Mixins Nocivos"
author: [gaearon]
```

Author: ahhhh puts kkk nem tinha percebido que era isso. Valeu!

Para facilitar a adoção e o aprendizado inicial, incluimos algumas válvulas de escape no React. O sistema mixin era uma dessas válvulas de escape, e seu objetivo era dar a você uma maneira de reutilizar o código entre os componentes, quando você não tem certeza de como resolver o mesmo problema com a composição.

Três anos se passaram desde que o React foi lancado. O cenário mudou. Agora, várias bibliotecas de visualizões adotam um modelo de componente semelhante ao React. Usar composição sobre a herança para criar interfaces de usuário declarativas não é mais uma novidade. Também estamos mais confiantes no modelo do componente React, e vimos muitos usos criativos dele tanto internamente quanto na comunidade.

Neste post, vamos considerar os problemas comumentes causados por mixins. Em seguida, sugeriremos vários padrões alternativos para os mesmos casos de uso. Descobrimos que esses padrões escalam melhor com a complexidade da base de código do que os mixins.

> ## Porque Mixins estão quebrados  {#porque-mixins-estao-quebrados}

Reviewer 2: Observando os outros documentos traduzidos é possível perceber que as âncoras/links não são modificados.
```suggestion
## Por que Mixins estão quebrados?  {#why-mixins-are-broken}
```

No entando, é inevitável que parte do nosso código usando React se torne gradualmente incompreensível. Ocasionalmente, a equipe do React veria grupos de componentes em diferentes projetos que as pessoas tinham medo de tocar. Esses componentes eram muito fáceis de serem quebrados acidentalmente, eram confusos para novos desenvolvedores e acabaram se tornando tão confusos para as pessoas que os escreveram em primeiro lugar. Muito dessa confusão ocorreu por causa dos mixins. Na época, eu não estava trabalhando no Facebook, mas cheguei [as mesmas conclusões]((https://medium.com/@dan_abramov/mixins-are-dead-long-live-higher-order-components-94a0d2f9e750)) depois de escrever vários de terríveis mixins.

Isso não significa que os próprios mixins são ruins. As pessoas empregam com sucesso em diferentes linguagens e paradigmas, incluindo algumas linguagens funcionais. No Facebook, usamos extensivamente traços no Hack que são bastante semelhantes aos mixins. No entanto, pensamos que os mixins são desnecessários e problemáticos nas bases de código do React. E aqui está o porquê.

> ### Mixins introduzem dependências implícitas {#mixins-introduzem-dependencias-implicitas}

Reviewer 2: ```suggestion
### Mixins introduzem dependências implícitas {#mixins-introduce-implicit-dependencies}
```

Essas dependências implícitas dificultam que novos membros da equipe contribuam para uma base de código. O método `render()` de um componente pode fazer referência a algum método que não está definido na classe. É seguro remover? Talvez esteja definido em um dos mixins. Mas qual deles? Você precisa rolar até a lista de mixins, abrir cada um desses arquivos e procurar por este método. Pior ainda, mixins podem especificar seus próprios mixins, então a pesquisa por se aprofundar.

Muitas vezes, os mixins passam a depender de outros mixins, e a remoção de um deles quebra o outro. Nessas situações, é muito complicado dizer como os dados entram e saem dos mixins e como é o seu gráfico de dependências. Ao contrário dos componentes, os mixins não formam uma hierarquia: eles são achatados e operam no mesmo namespace.

> ### Mixins causam confronto de nomes {#mixins-causam-confronto-de-nomes}

Reviewer 2: ```suggestion
### Mixins causam confronto de nomes {#mixins-cause-name-clashes}
```

> Não há garantia de que duas misturas específicas possam ser usadas juntas. Por exemplo, se `FluxListenerMixin` define `handleChange()` e `WindowSizeMixin` define `handleChange()`, voceê não pode usá-los juntos. Você também não pode definir um método com esse nome em seu próprio componente.

Reviewer 2: ```suggestion
Não há garantia de que duas misturas específicas possam ser usadas juntas. Por exemplo, se `FluxListenerMixin` define `handleChange()` e `WindowSizeMixin` define `handleChange()`, você não pode usá-los juntos. Você também não pode definir um método com esse nome em seu próprio componente.
```

Não é um grande problema se você controlar o código do <em>mixin</em>. Quando você tem um conflito, você pode renomear esse método em um dos <em>mixins</em>. No entanto, é complicado porque alguns componentes ou outros <em>mixins</em> já podem estar chamando esse métodos diretamente, e você precisa encontrar e corrigir essas chamadas também.

> Se você tiver um conflito de nome com um <em>mixin</em> de um pacote de terceiros, não será possível renomear um método dele. Em vez disso, você precisa usar nomes estanhos de métodos em seu componente para evitar conflitos.

Reviewer 2: ```suggestion
Se você tiver um conflito de nome com um <em>mixin</em> de um pacote de terceiros, não será possível renomear um método dele. Em vez disso, você precisa usar nomes desconhecidos de métodos em seu componente para evitar conflitos.
```

A situação não é melhor para os autores dos <em>mixins</em>. Até mesmo a adição de um novo método a um <em>mixin</em> é sempre uma alteração potencial, pois um método com o memso nome já pode existir em alguns dos componentes que o usam, diretamente ou por meio de outro <em>mixin</em>. Uma vez escritos, os <em>mixins</em> são difíceis de remover ou mudar. Idéias ruins não são refatoradas porque a refatoração é muito arriscada.

> ### Mixins desencadeiam complexidade de bola de neve {#mixins-desencadeiam-complexidade-de-bola-de-neve}

Reviewer 2: ```suggestion
### Mixins desencadeiam complexidade de bola de neve {#mixins-cause-snowballing-complexity}
```

Um componente precisa de algum estado para rastrear o foco do mouse. Para manter essa lógica reutilizável, você pode extrair `handleMouserEnter()`, `handleMouseLeave()` e `isHovering()` em um `HoverMixin`. Em seguida, alguém precisa implementar uma dica de ferramenta. Eles não querem duplicar a lógica em `HoverMixin`, de modo que criam um `TooltipMixin` que usa o `HoverMixin`. `TooltipMix` lê `isHovering()` fornecido pelo `HoverMixin` no seu `componentDidUpdate()` e mostra ou oculta a dica da ferramenta.

Alguns meses depois, alguém quer tornar a direção da dica de ferramenta configurável. Em um esforço para evitar a duplicação de código, eles adicionam suporte a um novo método opcional chamado `getTooltipOptions ()` em `TooltipMixin`. A essa altura, os componentes que mostram popovers também usam o `HoverMixin`. No entanto, os popovers precisam de um atraso diferente. Para resolver isso, alguém adiciona suporte para um método opcional `getHoverOptions ()` e o implementa no `TooltipMixin`. Esses mixins estão agora fortemente acoplados.

> Isso é bom enquanto não há novos requisitos. No entanto esta solução não escala muito bem. E se você deseja oferecer suporte à exibição de várias dicas de ferramentas em um único componente? Você não pode definir o mesmo mixin duas vezes em um componente. E se as dicas de ferramentas precisarem ser exibidas automaticamente em um tour guiado ao invés do houver? Boa sorte ao separar o `TooltipMixin` do` HoverMixin`. E se você precisar dar suporte ao caso onde a area do hover e a âncora da dica de ferramenta estejam localizadas em componentes diferentes? Você não pode elevar facilmente o estado usado misturando-se ao componente pai. Ao contrário dos componentes, os mixins não se servem para essas mudanças.

Reviewer 2: ```suggestion
Isso é bom enquanto não há novos requisitos. No entanto esta solução não escala muito bem. E se você deseja oferecer suporte à exibição de várias dicas de ferramentas em um único componente? Você não pode definir o mesmo mixin duas vezes em um componente. E se as dicas de ferramentas precisarem ser exibidas automaticamente em um tour guiado ao invés do hover? Boa sorte ao separar o `TooltipMixin` do` HoverMixin`. E se você precisar dar suporte ao caso onde a área do hover e a âncora da dica de ferramenta estejam localizadas em componentes diferentes? Você não pode elevar facilmente o estado usado misturando-se ao componente pai. Ao contrário dos componentes, mixins não se prestam naturalmente a essas mudanças.
```

Cada novo requisito torna os mixins mais difíceis de entender. Os componentes que usam o mesmo mixin ficam cada vez mais acoplados ao tempo. Qualquer novo recurso é adicionado a todos os componentes usando esse mixin. Não há como dividir uma parte "mais simples" do mixin sem duplicar o código ou introduzir mais dependências e indiretas entre mixins. Gradualmente, os limites do encapsulamento diminuem e, como é difícil alterar ou remover os mixins existentes, eles ficam cada vez mais abstratos até que ninguém entenda como eles funcionam.

Esses são os mesmos problemas que enfrentamos ao criar aplicativos antes do React. Descobrimos que eles são resolvidos por renderização declarativa, fluxo de dados de cima para baixo e componentes encapsulados. No Facebook, estamos migrando nosso código para usar padrões alternativos para mixins, e geralmente estamos felizes com os resultados. Você pode ler sobre esses padrões logo abaixo.

> ## Migrando de Mixins {#migrando-de-mixins}

Reviewer 2: ```suggestion
## Migrando de Mixins {#migrating-from-mixins}
```

Vamos deixar claro que mixins não são tecnicamente obsoletos. Se você usar `React.createClass ()`, poderá continuar usando-os. Dizemos apenas que eles não funcionaram bem para nós e, portanto, não recomendamos usá-los no futuro.

Cada seção abaixo corresponde a um padrão de uso de mixin que encontramos na base de código do Facebook. Para cada um deles, descrevemos o problema e uma solução que achamos que funciona melhor que os mixins. Os exemplos estão escritos no ES5, mas quando você não precisar de mixins, poderá mudar para as aulas do ES6, se desejar.

Esperamos que você ache esta lista útil. Informe-nos se não citamos algum caso de uso importante, para que possamos alterar a lista ou provar que estamos errados!

> ### Otimizações de desempenho {#otimizacao-de-desempenho}

Reviewer 2: ```suggestion
### Otimizações de desempenho {#performance-optimizations}
```

> Uma das mixins mais usadas é o [`PureRenderMixin`](/docs/pure-render-mixin.html). Você pode usá-lo em alguns componentes para [impedir repetições desnecessárias](/docs/advanced-performance.html#shouldcomponentupdate-in-action) quando os props e o estado são superficialmente iguais aos props e estado anteriores:

Reviewer 2: ```suggestion
Um dos mixins mais usados é o [`PureRenderMixin`](/docs/pure-render-mixin.html). Você pode usá-lo em alguns componentes para [impedir repetições desnecessárias](/docs/advanced-performance.html#shouldcomponentupdate-in-action) quando os props e o estado são superficialmente iguais aos props e estado anteriores:
```

});
```

> #### Solução {#solucao}

Reviewer 2: ```suggestion
#### Solução {#solution}
```

});
```
Se você usar um mix personalizado implementando uma função `shouldComponentUpdate` com algoritmo diferente, sugerimos exportar apenas essa função única de um módulo e chamá-la diretamente de seus componentes.

> Entendemos que mais digitação pode ser chato. Para o caso mais comuns, planejamos [introduzir uma nova classe base](https://github.com/facebook/react/pull/7195) chamada `React.PureComponent` na próxima versão menor. Ele usa a mesma comparação superficial que o `PureRenderMixin` faz hoje.

Reviewer 2: ```suggestion
Entendemos que mais digitação pode ser chato. Para o caso mais comum, planejamos [introduzir uma nova classe base](https://github.com/facebook/react/pull/7195) chamada `React.PureComponent` na próxima versão menor. Ele usa a mesma comparação superficial que o `PureRenderMixin` faz hoje.
```

> ## Assinaturas e efeitos colaterais {#assinaturas-e-efeitos-colaterais}

Reviewer 2: ```suggestion
## Assinaturas e efeitos colaterais {#subscriptions-and-side-effects}
```

module.exports = CommentList;
```

> #### Solução {#solucao-1}

Reviewer 2: ```suggestion
#### Solução {#solution-1}
```

Se houver apenas um componente inscrito nessa fonte de dados, não há problema em incorporar a lógica de assinatura diretamente no componente. Evite abstrações prematuras.

Se vários componentes usaram esse mixin para assinar uma fonte de dados, uma boa maneira de evitar a repetição é usar um padrão chamado ["componentes de ordem superior"](https://medium.com/@dan_abramov/mixins-are-dead-long-live-high-order-components-94a0d2f9e750). Pode parecer intimidador, portanto, examinaremos mais de perto como esse padrão emerge naturalmente do modelo de componentes.

> #### Explicação dos componentes de ordem superior {# explicacao-dos-componentes-de-ordem-superior}

Reviewer 2: ```suggestion
#### Explicação dos componentes de ordem superior {#higher-order-components-explained}
```

}
```

Essas duas funções não são muito úteis, mas nos ajudam a demonstrar um padrão que podemos aplicar posteriormente aos componentes.

Digamos que queremos extrair a lógica de log dessas funções sem alterar suas assinaturas. Como podemos fazer isso? Uma solução elegante é escrever uma [função de ordem superior](https://en.wikipedia.org/wiki/Higher-order_function), ou seja, uma função que assume uma função como argumento e retorna uma função.

Novamente, parece mais intimidador do que realmente é:

> 

Reviewer: ```suggestion
```

```js
function withLogging(wrappedFunction) {
  // Retorna uma função com a mesma API...
  return function(x, y) {
    // ... que chama a função original
    var result = wrappedFunction(x, y);
    // ... mas também registra o resultado
    console.log('result:', result);
    return result;
  };
}
```

> A função de ordem superior `withLogging` nos permite escrever `add` e `multiply` sem as instruções de log e depois envolvê-las para obter `addAndLog` e `multiplyAndLog` com exatamente as mesmas assinaturas de antes:

Reviewer: ```suggestion

A função de ordem superior `withLogging` nos permite escrever `add` e `multiply` sem as instruções de log e depois envolvê-las para obter `addAndLog` e `multiplyAndLog` com exatamente as mesmas assinaturas de antes:
```

  },

  render: function() {
    // Passamos o estado atual como props para CommentList.
    return <CommentList comments={this.state.comments} />;
  }
});

module.exports = CommentListWithSubscription;
```

Há apenas um passo final a ser feito.

> Lembra como fizemos com `withLogging ()` pegar uma função e retornar outra função envolvendo-a? Podemos aplicar um padrão semelhante aos componentes React.

Reviewer: ```suggestion
Lembra como fizemos com `withLogging()` pegar uma função e retornar outra função envolvendo-a? Podemos aplicar um padrão semelhante aos componentes React.
```

> Escreveremos uma nova função chamada `withSubscription (WrappedComponent)`. Seu argumento pode ser qualquer componente React. Passaremos `CommentList` como `WrappedComponent`, mas também poderíamos aplicar `withSubscription ()` a qualquer outro componente em nossa base de código.

Reviewer: ```suggestion
Escreveremos uma nova função chamada `withSubscription(WrappedComponent)`. Seu argumento pode ser qualquer componente React. Passaremos `CommentList` como `WrappedComponent`, mas também poderíamos aplicar `withSubscription()` a qualquer outro componente em nossa base de código.
```

Esta função retornaria outro componente. O componente retornado gerenciaria a assinatura e renderizaria `<WrappedComponent />` com os dados atuais.

Chamamos esse padrão de "componente de ordem superior".

> A composição acontece no nível de renderização React, e não com uma chamada direta da função. É por isso que não importa se o componente agrupado está definido com `createClass ()`, como uma classe ou função ES6. Se `WrappedComponent` for um componente React, o componente criado por `withSubscription () `poderá renderizá-lo.

Reviewer: ```suggestion
A composição acontece no nível de renderização React, e não com uma chamada direta da função. É por isso que não importa se o componente agrupado está definido com `createClass()`, como uma classe ou função ES6. Se `WrappedComponent` for um componente React, o componente criado por `withSubscription()` poderá renderizá-lo.
```

  }
});

// withSubscription() retorna um novo componente que
// é inscrito na fonte de dados e renderiza
// <CommentList /> com os dados atualizados.
var CommentListWithSubscription = withSubscription(CommentList);

// O restante da aplicação está interessada no componente inscrito
// então exportamos em vez do CommentList original.
module.exports = CommentListWithSubscription;
```

#### Solution, Revisited {#solution-revisited}

> Agora que entendemos melhor os componentes de ordem superior, vamos dar uma olhada na solução completa que não envolve mixins. Existem algumas pequenas alterações anotadas com comentários nas entre linhas:

Reviewer 2: ```suggestion
Agora que entendemos melhor os componentes de ordem superior, vamos dar uma olhada na solução completa que não envolve mixins. Existem algumas pequenas alterações anotadas com comentários inseridos nas linhas:
```

  )
}

// Em vez de declarar CommentListWithSubscription,
// exportamos o componente envolvido
module.exports = withSubscription(CommentList);
```

Os componentes de ordem superior são um padrão poderoso. Você pode passar argumentos adicionais para eles, se desejar personalizar ainda mais o comportamento deles. Afinal, eles nem são um recurso do React. São apenas funções que recebem componentes e retornam componentes que os envolvem.

Como qualquer solução, os componentes de ordem superior têm suas próprias armadilhas. Por exemplo, se você usar fortemente [refs](/docs/more-about-refs.html), poderá notar que agrupar algo em um componente de ordem superior altera o ref para apontar para o componente de agrupamento. Na prática, desencorajamos o uso de referências para a comunicação de componentes, por isso não achamos que seja um grande problema. No futuro, podemos considerar a adição de [ref forwarding](https://github.com/facebook/react/issues/4213) ao React para resolver esse incômodo.

> ### Lógica de renderização {#logica-de-renderizacao}

Reviewer 2: ```suggestion
### Lógica de renderização {#rendering-logic}
```

var UserRow = React.createClass({
  mixins: [RowMixin],

  // Chamado por RowMixin.renderHeader()
  getHeaderText: function() {
    return this.props.user.fullName;
  },

  render: function() {
    return (
      <div>
        {this.renderHeader() /* Definido por RowMixin */}
        <h2>{this.props.user.biography}</h2>
      </div>
    )
  }
});
```

> Vários componentes podem estar compartilhando `RowMixin` para renderizar o cabeçalho, e cada um deles precisaria definir` getHeaderText () `.

Reviewer: ```suggestion
Vários componentes podem estar compartilhando `RowMixin` para renderizar o cabeçalho, e cada um deles precisaria definir `getHeaderText()`.
```

> #### Solução {# solução-2}

Reviewer: ```suggestion
#### Solução {#solution-2}
```

Se você ver a lógica de renderização dentro de um mixin, é hora de extrair um componente!

> Em vez de `RowMixin`, definiremos um componente `<RowHeader>`. Também substituiremos a convenção de definir um método `getHeaderText ()` pelo mecanismo padrão do fluxo de dados principais no React: passando adereços.

Reviewer: ```suggestion
Em vez de `RowMixin`, definiremos um componente `<RowHeader>`. Também substituiremos a convenção de definir um método `getHeaderText()` pelo mecanismo padrão do fluxo de dados principais no React: passando adereços.
```

  );
}
```
As props mantêm as dependências de componentes explícitas, fáceis de substituir e aplicáveis com ferramentas como [Flow](https://flowtype.org/) e [TypeScript](https://www.typescriptlang.org/).

> **Nota:**
>
> Definir componentes como funções não é necessário. Também não há nada de errado em usar métodos e estados do ciclo de vida - eles são os primeiros recursos do React quando usado em formato de classe. Usamos componentes de função neste exemplo porque são mais fáceis de ler e não precisávamos desses recursos extras, mas as classes funcionariam da mesma maneira.

> ### Contexto {#contexto}

Reviewer 2: ```suggestion
### Contexto {#context}
```

module.exports = Link;
```

> #### Solução {# solution-3}

Reviewer 2: ```suggestion
#### Solução {#solution-3}
```

> Concordamos que ocultar o uso do contexto do consumo de componentes é uma boa idéia até que a API do contexto se estabilize. No entanto, recomendamos o uso de componentes de ordem superior em vez de mixins para isso.

Reviewer 2: ```suggestion
Concordamos que ocultar o uso do contexto do consumo de componentes é uma boa ideia até que a API do contexto se estabilize. No entanto, recomendamos o uso de componentes de ordem superior em vez de mixins para isso.
```

  }
});

// Não se esqueça de empacotar o componente!
module.exports = withRouter(Link);
```

Se você estiver usando uma biblioteca de terceiros que fornece apenas um mixin, recomendamos que você arquive um problema vinculado a esta postagem para que eles possam fornecer um componente de ordem superior. Enquanto isso, você pode criar um componente de ordem superior exatamente da mesma maneira.

> ### Métodos de utilidade {#metodos-de-utilidade}

Reviewer 2: ```suggestion
### Métodos de utilidade {#utility-methods}
```

});
```

> #### Solução {# solution-4}

Reviewer 2: ```suggestion
#### Solução {#solution-4}
```

});
```

> ### Outros casos de uso {#outros-casos-de-uso}

Reviewer 2: ```suggestion
### Outros casos de uso {#other-use-cases}
```

Às vezes, as pessoas usam mixins para adicionar seletivamente o log aos métodos do ciclo de vida em alguns componentes. No futuro, pretendemos fornecer uma [API oficial do DevTools](https://github.com/facebook/react/issues/5306) que permita implementar algo semelhante sem tocar nos componentes. No entanto, ainda é um trabalho em andamento. Se você depende muito de registrar mixins para depuração, convém continuar usando esses mixins por mais algum tempo.

Se você não conseguir realizar algo com um componente, um componente de ordem superior ou um módulo utilitário, isso pode significar que o React deve fornecer isso imediatamente. [Arquive um problema](https://github.com/facebook/react/issues/new) para nos informar sobre seu caso de uso para mixins, e ajudaremos você a considerar alternativas ou talvez implementar sua solicitação de recurso.

> Mixins não são descontinuados no sentido tradicional. Você pode continuar usando-os com `React.createClass ()`, pois não mudaremos mais. Eventualmente, à medida que as classes ES6 ganham mais adoção e seus problemas de usabilidade no React são resolvidos, podemos dividir `React.createClass ()` em um pacote separado, porque a maioria das pessoas não precisa disso. Mesmo nesse caso, seus antigos mixins continuariam funcionando.

Reviewer: ```suggestion
Mixins não são descontinuados no sentido tradicional. Você pode continuar usando-os com `React.createClass()`, pois não mudaremos mais. Eventualmente, à medida que as classes ES6 ganham mais adoção e seus problemas de usabilidade no React são resolvidos, podemos dividir `React.createClass()` em um pacote separado, porque a maioria das pessoas não precisa disso. Mesmo nesse caso, seus antigos mixins continuariam funcionando.
```
